src/cocotb/_task_manager.py

```diff
@@ -0,0 +1,387 @@
+# Copyright cocotb contributors
+# Licensed under the Revised BSD License, see LICENSE for details.
+# SPDX-License-Identifier: BSD-3-Clause
+
+"""TaskManager and related code."""
+
+from __future__ import annotations
+
+import sys
+from asyncio import CancelledError
+from bdb import BdbQuit
+from types import CoroutineType
+from typing import TYPE_CHECKING, Any, Literal, TypeVar
+
+import cocotb
+from cocotb._base_triggers import NullTrigger
+from cocotb.task import Task, current_task
+from cocotb.triggers import Event, Trigger
+
+if TYPE_CHECKING:
+    from collections.abc import Awaitable, Callable, Coroutine
+
+if sys.version_info >= (3, 11):
+    from typing import Self
+else:
+    from exceptiongroup import BaseExceptionGroup
+
+
+T = TypeVar("T")
+
+
+async def _waiter(aw: Awaitable[T]) -> T:
+    return await aw
+
+
+class TaskManager:
+    r"""An :term:`asynchronous context manager` which enables the user to run :term:`coroutine function`\ s or :keyword:`await` :term:`awaitable`\ s concurrently and wait for them all to finish.
+
+    The :deco:`fork` decorator can be used to run :term:`coroutine function`\ s in parallel as children of the :class:`!TaskManager` object.
+    When control reaches the end of the context block
+    the :class:`!TaskManager` blocks the calling :class:`!Task` until all children :class:`!Task`\ s complete.
+
+    .. code-block:: python
+
+        async with TaskManager() as tm:
+
+            @tm.fork
+            async def drive_if1(): ...
+
+            @tm.fork
+            async def drive_if2(): ...
+
+        # Control returns here when all drive Tasks have completed
+
+    After exiting the context block,
+    if any child :class:`!Task` fails with an exception,
+    one of two behaviors will occur:
+    if the ``continue_on_error`` parameter is ``False`` (default), all other child :class:`!Task`\ s are cancelled;
+    if the ``continue_on_error`` parameter is ``True``, other child :class:`!Task`\ s are allowed to continue running.
+
+    After all child :class:`!Task`\ s have finished,
+    all exceptions, besides :exc:`CancelledError`, are gathered into an :exc:`ExceptionGroup`,
+    or an :exc:`BaseExceptionGroup`, if at least one of the exceptions is a :exc:`BaseException`,
+    and raised in the enclosing scope.
+
+    You can catch the :exc:`!ExceptionGroup` to handle errors from child :class:`!Task`\ s
+    by either catching the :exc:`!ExceptionGroup` as you would typically;
+    or, if you are running Python 3.11 or later,
+    using the new ``except*`` syntax to catch specific exception types from the group.
```

Comment:
Would be nice to link to https://docs.python.org/3/reference/compound_stmts.html#except-star